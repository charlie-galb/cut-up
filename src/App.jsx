import React from 'react'

const App = () => {
  return (
    <h1 data-testid="test-header">
      Cut-up App
    </h1>
  )
}

export default App