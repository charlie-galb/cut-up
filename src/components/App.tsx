import React, { useState } from "react"

import { CuttingBoard } from "./CuttingBoard/CuttingBoard"
import { PasteBoard } from "./PasteBoard/PasteBoard"

export const App = () => {

  const [snippedText, setSnippedText] = useState<string[]>([""])

  const snipText = (text: string) => {
    const temp = text.split(" ")
    const result: Array<string> = []
    for(let i = 0; i < temp.length; i = i + 2 ) {
      result.push(temp.slice(i,i+2).join(' '))
    }
    setSnippedText(result)
  }

  return (
    <div>
      <h1 data-testid="test-header">Cut-up App</h1>
      <CuttingBoard snipText={snipText}/>
      <PasteBoard wordChunks={snippedText} />
    </div>
    
  )
}
