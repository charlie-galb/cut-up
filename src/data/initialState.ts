import { chunk } from '../types/chunk'
import { chunkContainer } from '../types/chunkContainer'

interface initialState {
    activeId: string
    overId: string
    wordChunks: {
      [key: string]: string
    }
    chunkContainers: {
      [key: string]: chunkContainer
    }
    lineOrder: string[],
    poemAsText: string
  }

  const pasteBoard: chunkContainer = {
    id: 'chunk-container-1',
    title: 'pasteboard',
    nestedChunkIDs: []
  }

  const line1: chunkContainer = {
    id: 'chunk-container-2',
    title: 'line-1',
    nestedChunkIDs: []
  }

export const initialState: initialState = {
    activeId: "",
    overId: "",
    wordChunks: {},
    chunkContainers: {
      'chunk-container-1': pasteBoard,
      'chunk-container-2': line1
    },
    lineOrder: ['chunk-container-2'],
    poemAsText: ""
}