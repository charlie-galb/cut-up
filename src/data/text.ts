export const text = {
    para1: "Cut Up Creator is a tool for making collage poems in the style of William S. Burroughs.",
    para2: "To begin, just copy text from three different sources (the more different the better) and paste them in the boxes below. Then hit the cut-up button to shred the text.",
    para3: "Drag and drop the snippets into the blank lines provided, creating whatever weird combinations you like. When you're happy with what you've made, hit the textify button. Your poem will appear as text below.",
}