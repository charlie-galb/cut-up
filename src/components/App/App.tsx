import React, { useState } from "react"

import { DndContext, useSensor, useSensors, KeyboardSensor, MouseSensor, TouchSensor, } from '@dnd-kit/core'

import { CuttingBoard } from "../CuttingBoard/CuttingBoard"
import { CraftingBoard } from "../CraftingBoard/CraftingBoard"
import { Header } from "../Header/Header"

import { initialChunkContainers, initialLineOrder } from "../../data/initialState"
import { text } from "../../data/text"

import { removeAtIndex, insertAtIndex, arrayMove } from "../../utils/array"

import { chunkContainer } from "../../types/chunkContainer"


export const App = () => {
  const [activeId, setActiveId] = useState<string>("")
  const [wordChunks, setWordChunks] = useState< { [key: string]: string } >({})
  const [chunkContainers, setChunkContainers] = useState<{ [key: string]: chunkContainer}>(initialChunkContainers)
  const [lineOrder, setLineOrder] = useState<string[]>(initialLineOrder)
  const [poemAsText, setPoemAsText] = useState<string>("")

  const sensors = useSensors(
    useSensor(MouseSensor),
    useSensor(TouchSensor),
    useSensor(KeyboardSensor),
  );

  const { para1, para2, para3 } = text

  const onDragStart = (event: any) => {
    const { active } = event
    const { id } = active
    setActiveId(id)
  }

  const onDragCancel = (event: any) => {
    setActiveId("")
  }

  const onDragOver = (event: any) => {
    const { active, over } = event
    const overId = over?.id

    if (!overId) { return }

    const activeContainerId: string = active.data.current.sortable.containerId
    const activeContainer: chunkContainer = chunkContainers[activeContainerId]
    const overContainerId: string = over.data.current?.sortable.containerId || over.id
    const overContainer: chunkContainer = chunkContainers[overContainerId]

    if (!overContainer) { return }

    if (activeContainerId !== overContainerId) {

      const activeChunks = activeContainer.nestedChunkIDs
      const overChunks = overContainer.nestedChunkIDs
      const activeIndex = activeChunks.indexOf(active.id)
      const overIndex = 
        over.id in chunkContainers ? overChunks.length + 1 : overChunks.indexOf(over.id)
      const draggedItem = active.id
      
      setChunkContainers(moveBetweenContainers(
        activeChunks,
        activeContainer, 
        activeIndex, 
        overChunks, 
        overContainer, 
        overIndex,
        draggedItem
      ))
    }
  }

  const onDragEnd = (event: any) => {
    const {active, over} = event
    
    if (!over || !active) { 
      setActiveId("")
      return 
    }

    const activeContainerId = active.data.current.sortable.containerId
    const activeContainer = chunkContainers[activeContainerId]
    const overContainerId = over.data.current?.sortable.containerId || over.id
    const overContainer = chunkContainers[overContainerId]
   
    if (active.id !== over.id && activeContainerId === overContainerId) {

      const items = activeContainer.nestedChunkIDs
      const activeIndex = items.indexOf(active.id);
      const overIndex = items.indexOf(over.id);
      const sortedItems = arrayMove(items, activeIndex, overIndex)

      setChunkContainers({
        ...chunkContainers,
          [activeContainerId]: {
            ...activeContainer,
            nestedChunkIDs: sortedItems
          }
        })
    } 

    if (activeContainerId !== overContainerId) {

      const activeChunks = activeContainer.nestedChunkIDs
      const overChunks = overContainer.nestedChunkIDs
      const activeIndex = activeChunks.indexOf(active.id)
      const overIndex = overChunks.indexOf(over.id)
      const draggedItem = active.id
      
      setChunkContainers(moveBetweenContainers(
        activeChunks,
        activeContainer, 
        activeIndex, 
        overChunks, 
        overContainer, 
        overIndex,
        draggedItem
      ))
    }
    setActiveId("")
  }

  const moveBetweenContainers = (
    activeChunks: string[],
    activeContainer: chunkContainer,
    activeIndex: number,
    overChunks: string[],
    overContainer: chunkContainer,
    overIndex: number,
    item: string
  ) => {
    return {
      ...chunkContainers,
        [activeContainer.id]: {
          ...activeContainer,
          nestedChunkIDs: removeAtIndex(activeChunks, activeIndex)
        },
        [overContainer.id]: {
          ...overContainer,
          nestedChunkIDs: insertAtIndex(overChunks, overIndex, item)
        }
      }
    }

    return (
      <div className="app-container">
        <Header />
        <p data-testid='para-1'><b>{para1}</b></p>
        <p data-testid='para-2'>{para2}</p>
        <CuttingBoard 
          chunkContainers={chunkContainers}
          setWordChunks={setWordChunks}
          setChunkContainers={setChunkContainers}/>
        <p data-testid='para-3'>{para3}</p>
        <DndContext 
          sensors={sensors}
          onDragStart={onDragStart}  
          onDragCancel={onDragCancel}
          onDragOver={onDragOver} 
          onDragEnd={onDragEnd} >
          <CraftingBoard 
            activeId={activeId} 
            wordChunks={wordChunks} 
            chunkContainers={chunkContainers} 
            lineOrder={lineOrder} 
            setChunkContainers={setChunkContainers}
            setLineOrder={setLineOrder}
            setPoemAsText={setPoemAsText}
            poemAsText={poemAsText} />
        </DndContext>
      </div>
    )

  }

export default App