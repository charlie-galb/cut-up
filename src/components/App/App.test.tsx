import React from 'react';
import { render } from "@testing-library/react";

import { text } from '../../data/text'
import { App } from './App';

const { para1, para2, para3 } = text

describe('App', () => {
    test('renders without crashing', () => {
        const app = render(<App />);
        expect(app.getByTestId('para-1')).toHaveTextContent(para1) 
        expect(app.getByTestId('para-2')).toHaveTextContent(para2) 
        expect(app.getByTestId('para-3')).toHaveTextContent(para3) 
        expect(app).toMatchSnapshot();
    });
})
    