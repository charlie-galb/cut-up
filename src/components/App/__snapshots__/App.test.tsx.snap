// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders without crashing 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="app-container"
      >
        <div
          class="header"
        >
          <h1
            data-testid="header"
          >
            Cut-up Creator
          </h1>
        </div>
        <p
          data-testid="para-1"
        >
          <b>
            Cut Up Creator is a tool for making collage poems in the style of William S. Burroughs.
          </b>
        </p>
        <p
          data-testid="para-2"
        >
          To begin, just copy text from three different sources (the more different the better) and paste them in the boxes below. Then hit the cut-up button to shred the text.
        </p>
        <div
          class="cutting-container"
        >
          <div
            class="text-area-container"
          >
            <textarea
              class="cutting-text-area 1"
              cols="10"
              data-testid="cutting-text-area1"
              placeholder="Paste your text sources here"
              rows="4"
            />
            <textarea
              class="cutting-text-area 2"
              cols="10"
              data-testid="cutting-text-area2"
              placeholder="Paste your text sources here"
              rows="4"
            />
            <textarea
              class="cutting-text-area 3"
              cols="10"
              data-testid="cutting-text-area3"
              placeholder="Paste your text sources here"
              rows="4"
            />
          </div>
          <button
            class="btn primary-btn cut-btn"
            data-testid="cut-btn"
          >
            Cut up
          </button>
        </div>
        <p
          data-testid="para-3"
        >
          Drag and drop the snippets into the blank lines provided, creating whatever weird combinations you like. When you're happy with what you've made, hit the textify button. Your poem will appear as text below.
        </p>
        <div
          class="crafting-background"
        >
          <div
            class="crafting-container"
          >
            <div
              class="pasteboard-container"
            >
              <div>
                <div
                  class="unused-snippets"
                  data-rbd-droppable-context-id="0"
                  data-rbd-droppable-id="chunk-container-1"
                  data-testid="pasteboard"
                />
              </div>
            </div>
            <div
              class="wip-container"
            >
              <div
                class="lines-container"
              >
                <div>
                  <div
                    class="line"
                    data-rbd-droppable-context-id="0"
                    data-rbd-droppable-id="chunk-container-2"
                    data-testid="line-1"
                  />
                </div>
              </div>
              <div>
                <button
                  class="btn secondary-btn add-line-btn"
                  data-testid="add-line-btn"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    />
                  </svg>
                </button>
              </div>
              <div>
                <button
                  class="btn secondary-btn remove-line-btn"
                  data-testid="remove-line-btn"
                >
                  <svg
                    fill="currentColor"
                    height="1em"
                    stroke="currentColor"
                    stroke-width="0"
                    viewBox="0 0 448 512"
                    width="1em"
                    xmlns="http://www.w3.org/2000/svg"
                  >
                    <path
                      d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </div>
          <div
            class="textify-button-container"
            data-testid="textify-button-container"
          >
            <button
              class="btn primary-btn textify-btn"
              data-testid="textify-btn"
            >
              Textify
            </button>
          </div>
        </div>
      </div>
    </div>
    <div
      aria-atomic="true"
      aria-live="assertive"
      id="rbd-announcement-0"
      style="position: absolute; width: 1px; height: 1px; margin: -1px; border: 0px; padding: 0px; overflow: hidden; clip-path: inset(100%);"
    />
    <div
      id="rbd-hidden-text-0-hidden-text-0"
      style="display: none;"
    >
      
  Press space bar to start a drag.
  When dragging you can use the arrow keys to move the item around and escape to cancel.
  Some screen readers may require you to be in focus mode or to use your pass through key

    </div>
  </body>,
  "container": <div>
    <div
      class="app-container"
    >
      <div
        class="header"
      >
        <h1
          data-testid="header"
        >
          Cut-up Creator
        </h1>
      </div>
      <p
        data-testid="para-1"
      >
        <b>
          Cut Up Creator is a tool for making collage poems in the style of William S. Burroughs.
        </b>
      </p>
      <p
        data-testid="para-2"
      >
        To begin, just copy text from three different sources (the more different the better) and paste them in the boxes below. Then hit the cut-up button to shred the text.
      </p>
      <div
        class="cutting-container"
      >
        <div
          class="text-area-container"
        >
          <textarea
            class="cutting-text-area 1"
            cols="10"
            data-testid="cutting-text-area1"
            placeholder="Paste your text sources here"
            rows="4"
          />
          <textarea
            class="cutting-text-area 2"
            cols="10"
            data-testid="cutting-text-area2"
            placeholder="Paste your text sources here"
            rows="4"
          />
          <textarea
            class="cutting-text-area 3"
            cols="10"
            data-testid="cutting-text-area3"
            placeholder="Paste your text sources here"
            rows="4"
          />
        </div>
        <button
          class="btn primary-btn cut-btn"
          data-testid="cut-btn"
        >
          Cut up
        </button>
      </div>
      <p
        data-testid="para-3"
      >
        Drag and drop the snippets into the blank lines provided, creating whatever weird combinations you like. When you're happy with what you've made, hit the textify button. Your poem will appear as text below.
      </p>
      <div
        class="crafting-background"
      >
        <div
          class="crafting-container"
        >
          <div
            class="pasteboard-container"
          >
            <div>
              <div
                class="unused-snippets"
                data-rbd-droppable-context-id="0"
                data-rbd-droppable-id="chunk-container-1"
                data-testid="pasteboard"
              />
            </div>
          </div>
          <div
            class="wip-container"
          >
            <div
              class="lines-container"
            >
              <div>
                <div
                  class="line"
                  data-rbd-droppable-context-id="0"
                  data-rbd-droppable-id="chunk-container-2"
                  data-testid="line-1"
                />
              </div>
            </div>
            <div>
              <button
                class="btn secondary-btn add-line-btn"
                data-testid="add-line-btn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 208H272V64c0-17.67-14.33-32-32-32h-32c-17.67 0-32 14.33-32 32v144H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h144v144c0 17.67 14.33 32 32 32h32c17.67 0 32-14.33 32-32V304h144c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  />
                </svg>
              </button>
            </div>
            <div>
              <button
                class="btn secondary-btn remove-line-btn"
                data-testid="remove-line-btn"
              >
                <svg
                  fill="currentColor"
                  height="1em"
                  stroke="currentColor"
                  stroke-width="0"
                  viewBox="0 0 448 512"
                  width="1em"
                  xmlns="http://www.w3.org/2000/svg"
                >
                  <path
                    d="M416 208H32c-17.67 0-32 14.33-32 32v32c0 17.67 14.33 32 32 32h384c17.67 0 32-14.33 32-32v-32c0-17.67-14.33-32-32-32z"
                  />
                </svg>
              </button>
            </div>
          </div>
        </div>
        <div
          class="textify-button-container"
          data-testid="textify-button-container"
        >
          <button
            class="btn primary-btn textify-btn"
            data-testid="textify-btn"
          >
            Textify
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
